src/kinetics/Reaction.cpp

```diff
@@ -276,28 +276,49 @@ ThreeBodyReaction::ThreeBodyReaction(const Composition& reactants_,
 }
 
 std::string ThreeBodyReaction::reactantString() const {
-    return ElementaryReaction::reactantString() + " + M";
+    if (specified_collision_partner) {
+        return ElementaryReaction::reactantString() + " + "
+            + third_body.efficiencies.begin()->first;
+    } else {
+        return ElementaryReaction::reactantString() + " + M";
+    }
 }
 
 std::string ThreeBodyReaction::productString() const {
-    return ElementaryReaction::productString() + " + M";
+    if (specified_collision_partner) {
+        return ElementaryReaction::productString() + " + "
+            + third_body.efficiencies.begin()->first;
+    } else {
+        return ElementaryReaction::productString() + " + M";
+    }
 }
 
 void ThreeBodyReaction::calculateRateCoeffUnits(const Kinetics& kin)
 {
     ElementaryReaction::calculateRateCoeffUnits(kin);
-    const ThermoPhase& rxn_phase = kin.thermo(kin.reactionPhaseIndex());
-    rate_units *= rxn_phase.standardConcentrationUnits().pow(-1);
+    bool specified_collision_partner_ = false;
+    for (const auto& reac : reactants) {
+        if (reac.first != "M" && products.count(reac.first)) {
+            // detected specified third-body collision partner
+            specified_collision_partner_ = true;
+        }
+    }
+    if (!specified_collision_partner_) {
+        const ThermoPhase& rxn_phase = kin.thermo(kin.reactionPhaseIndex());
+        rate_units *= rxn_phase.standardConcentrationUnits().pow(-1);
+    }
```

Comment:
Sorry for not mentioning this before, but could you add some comments here to help explain the somewhat tricky behavior? I assume it all has to do with the fact that `calculateRateCoeffUnits` is often called before the rest of the setup for the specific reaction type, which would do things like remove `M` from the reactants and products list.

Reply:
No problem. The logic implemented here is indeed to catch these cases. I added the comment as requested.